files/class-vip-filesystem-stream-wrapper.php

```diff
@@ -854,13 +854,41 @@ public function validate( $path, $mode ) {
 	 * @since   1.0.0
 	 * @access  protected
 	 * @param   string    $message  Debug message to be logged
+	 * @param   bool      $force Whether to force debug
 	 */
-	protected function debug( $message ) {
-		if ( ! $this->debug_mode ) {
+	protected function debug( $message, $force = false ) {
+		if ( ! ( $this->debug_mode || $force ) ) {
 			return;
 		}
 
-		// phpcs:ignore WordPress.PHP.DevelopmentFunctions.error_log_error_log
-		error_log( sprintf( '%s (%s)', $message, wp_debug_backtrace_summary() ) );
+		$trace = $this->backtrace_fmt();
+
+		\Automattic\VIP\Logstash\log2logstash(
+			[
+				'severity' => 'info',
+				'feature'  => 'stream_wrapper_audit_' . $trace[1]['function'],
+				'message'  => "File op {$trace[1]['function']}: " . $message,
+				'extra'    => [
+					'trace' => $this->backtrace_fmt(),
+				],
+			]
+		);
+	}
+
+	/**
+	 * Format the debug backtrace to be a bit more readable .
+	 *
+	 * @return array
+	 */
+	private function backtrace_fmt() {
+		$trace = debug_backtrace( false, 30 ); // phpcs:ignore WordPress.PHP.DevelopmentFunctions.error_log_debug_backtrace
+		// Discard current frame.
+		unset( $trace[0] );
+		foreach ( $trace as &$frame ) {
+			$frame['file'] = str_replace( ABSPATH, '', $frame['file'] ) . ':' . $frame['line'];
+			unset( $frame['line'] );
```

Comment:
Maybe

```php
if ( isset( $frame['file'] ) ) {
    $frame['file'] = str_replace( ABSPATH, '', $frame['file'] ) . ':' . $frame['line'];
    unset( $frame['line'] );
}
```

as it is possible for the frame not to contain the file/line information (for example, when a function is called from a PHP function):

```php
function process() {
    print_r( debug_backtrace() );
}

array_map( 'process', [1] );
```

will result in

```
Array
(
    [0] => Array
        (
            [function] => process
            [args] => Array
                (
                    [0] => 1
                )

        )

    [1] => Array
        (
            [file] => test.php
            [line] => 7
            [function] => array_map
            [args] => Array
                (
                    [0] => process
                    [1] => Array
                        (
                            [0] => 1
                        )

                )

        )

)
```

My understanding is that wrapper's functions will be called from PHP's internal functions, and this is where we might find frames lacking file/line information.